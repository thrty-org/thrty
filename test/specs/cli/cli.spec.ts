import { spawnSync } from 'child_process';
import { SpawnSyncReturns } from 'node:child_process';
import { readdir } from 'node:fs/promises';
import { join } from 'path';
import { existsSync } from 'node:fs';
import { mkdirSync } from 'fs';

describe('thrty create-api-client', () => {
  let result: SpawnSyncReturns<Buffer>;

  describe.each([
    ['--httpClient', 'axios'],
    ['--httpClient', 'fetch'],
  ])('%s=%s', (httpClientFlag, httpClientOption) => {
    const outPath = join('dist', `${httpClientOption}TodoApi.ts`);

    beforeAll(() => {
      const outDir = join(__dirname, 'dist');
      if (!existsSync(outDir)) {
        mkdirSync(outDir, { recursive: true });
      }
      result = spawnSync('npx', [
        'thrty',
        'create-api-client',
        'specs/__fixtures__/*Lambda.ts',
        '--name=todoApi',
        `--outPath=specs/cli/${outPath}`,
        `${httpClientFlag}=${httpClientOption}`,
        '--models=zod',
      ]);
    });

    it('should not result in errors', () => {
      expect(result.stderr.toString()).toEqual('');
    });

    it('should create api client', async () => {
      const { todoApiFactory }: any = await import(`./${outPath}` as any);
      expect(todoApiFactory.toString()).toMatchSnapshot();
    });
  });
});

describe('thrty create-api-package', () => {
  let result: SpawnSyncReturns<Buffer>;

  const outDir = join('dist', `todo-api-generated`);

  beforeAll(() => {
    result = spawnSync('npx', [
      'thrty',
      'create-api-package',
      'todo-api-generated',
      'specs/__fixtures__/*Lambda.ts',
      '--name=todoApi',
      `--outDir=specs/cli/${outDir}`,
      '--httpClient=axios',
      '--models=zod',
    ]);
  });

  it('should not result in errors', () => {
    expect(result.stderr.toString()).toEqual('');
  });

  it('should create package with source files and folders', async () => {
    const _outDir = join(__dirname, outDir);
    const packagePathParts = await readdir(_outDir);
    expect(packagePathParts).toEqual(['package.json', 'src']);

    const srcPathParts = await readdir(join(_outDir, 'src'));
    expect(srcPathParts).toEqual(['index.ts']);
  });
});
