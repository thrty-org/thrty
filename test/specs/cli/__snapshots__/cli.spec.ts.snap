// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`thrty create-api-client --httpClient=axios should create api client 1`] = `
"({ httpClient }) => ({
    async patchTodo(todoId, data, options) {
        const path = '/todos/{todoId}'.replace('{todoId}', encodeURIComponent(String(todoId)));
        return httpClient(path, {
            method: 'patch',
            data,
            ...options,
        });
    },
    async getTodos(options) {
        const path = '/todos';
        return httpClient(path, {
            method: 'get',
            ...options,
        });
    },
    async createTodo(data, options) {
        const path = '/todos';
        return httpClient(path, {
            method: 'post',
            data,
            ...options,
        });
    },
})"
`;

exports[`thrty create-api-client --httpClient=fetch should create api client 1`] = `
"({ baseUrl, requestInterceptor }) => ({
    async patchTodo(todoId, data, options) {
        const path = '/todos/{todoId}'.replace('{todoId}', encodeURIComponent(String(todoId)));
        const request = { data: data, method: 'patch', baseUrl: baseUrl, path: path };
        return fetch(request.baseUrl + request.path, { ...request, ...(await requestInterceptor?.(request)), ...options }).then(async (res) => ({ data: await res.json(), ...res }));
    },
    async getTodos(options) {
        const path = '/todos';
        const request = { method: 'get', baseUrl: baseUrl, path: path };
        return fetch(request.baseUrl + request.path, { ...request, ...(await requestInterceptor?.(request)), ...options }).then(async (res) => ({ data: await res.json(), ...res }));
    },
    async createTodo(data, options) {
        const path = '/todos';
        const request = { data: data, method: 'post', baseUrl: baseUrl, path: path };
        return fetch(request.baseUrl + request.path, { ...request, ...(await requestInterceptor?.(request)), ...options }).then(async (res) => ({ data: await res.json(), ...res }));
    },
})"
`;
